### PR TITLE
Make the bit width configurable.

### PR DESCRIPTION
**Pull request by Stebalien**
NOTE: This also increases the max index to uint64-1. The previous max index was set to "max int" so the max index allowed us to nicely fill the max height. However, with arbitrary heights, that no longer makes any sense.

Unfortunately, our "array" abstraction uses int64, not uint64 (and our max sector number is int64). We'll need to audit specs-actors to make sure we're enforcing maximums in all the correct places. Alternatively, we could drop the max back down, but that would be unfortunate.

### CODE REVIEW

File: amt.go
Comment (Stebalien) [inline amt.go:34]: This is going to require changing all invocations of `MakeEmptyArray`. The alternative is to simply panic on bad options (which isn't totally unreasonable, but is a bit unfortunate).

File: node.go
Comment (Stebalien) [inline node.go:55]: I think this logic is correct, but please check it. It's non-trivial.

File: util.go
Comment (Stebalien) [inline util.go:12]: Please check this (happy to discuss but it would be nice to get independent confirmation).

File: util.go
Comment (Stebalien) [inline util.go:61]: "I'm sorry" doesn't quite cut it.

**Review by ZenGround0 — COMMENTED** [review]
Looked at everything except test and nodesForHeight and it looks good. Will try to get the rest by sometime tomorrow.

File: options.go
Comment (Stebalien) [reply options.go:20]: I could set an arbitrary one, but I don't see a reason to do so.

File: options.go
Comment (Stebalien) [reply options.go:17]: Done.

File: amt.go
Comment (Stebalien) [reply amt.go:63]: Specifically, it prevents `r.Height+1` from overflowing. I set it to 64 because nothing larger is sane.

Comment (Stebalien) [reply amt.go:63]: I'll improve the comment.

**Comment by anorth** [comment]
We can possibly change the array abstraction's index type. It's worth looking into to follow up.

File: amt.go
Comment (anorth) [reply amt.go:34]: Thanks. Absolutely please force all callers to acknowledge the change.

File: amt.go
Comment (anorth) [inline amt.go:43]: Can we store the bitWidth in the top node of the AMT? If that requires a new struct to do so, I'd say it's worth it. We should strive for as much self-description as we can in these structures.

File: go.mod
Comment (anorth) [inline go.mod:8]: This would be a really unfortunate dependency for this and, transitively, actors to have. And look at all the stuff it brings in! Happily, I can see nowhere where you've actually imported this so I think you can remove it.

If we do need some testing or benchmarking thing that needs large on-disk storage like this, I think we should pull it out to a separate repo that mere users of the AMT can avoid depending on. 

Blocking, if you can't remove it.

Comment (Stebalien) [reply go.mod]: I'm not sure how I ended up pulling that in. Go mod tidy took care of it.

File: amt.go
Comment (Stebalien) [reply amt.go:43]: Fair enough.

Comment (Stebalien) [reply amt.go:43]: @whyrusleeping as the author of this library, any thoughts here?

**Review by ZenGround0 — APPROVED** [review]

File: amt.go
Comment (rvagg) [reply amt.go:43]: 💯 please do this

Comment (ZenGround0) [reply amt.go:43]: @Stebalien are you blocking on this? If so is there a way I can help unblock?

Comment (Stebalien) [reply amt.go:43]: I'll ping why.

Comment (whyrusleeping) [reply amt.go:43]: eh, okay. we can do that now

Comment (Stebalien) [reply amt.go:43]: Ok, I've added it. However, I want to note that it does have limited usefulness:

* We can't rely on it in filecoin because we want to require a specific bitwidth. At best, we can use it in cross-version analysis tools.
* It won't help version across other changes like, e.g., fixing the AMT depth problem. That is, it'll make it easier for us to write tools that work across upgrades where we change the bitwidth, but not across upgrades where we change anything else.

Comment (anorth) [reply amt.go:43]: Thank you! I agree the benefit is greater for consumers outside the code that is actually writing the AMT, but I think that adds up to a lot over time (even though we won't directly learn of much of it).

**Comment by Stebalien** [comment]
@ZenGround0 can I get a final review on that last commit (encoding the bitwidth in the AMT)?

**Review by ZenGround0 — APPROVED** [review]

**Pull request merged**
